imports/lib/schemas/customTypes.ts

```diff
@@ -0,0 +1,43 @@
+import { Meteor } from 'meteor/meteor';
+import { z } from 'zod';
+import { Id } from './regexes';
+
+// Each of these is set to true based on the context in which the schema is
+// being evaluated. They are mutually exclusive (although they can all be false)
+export const IsInsert = new Meteor.EnvironmentVariable<boolean>();
+export const IsUpdate = new Meteor.EnvironmentVariable<boolean>();
+export const IsUpsert = new Meteor.EnvironmentVariable<boolean>();
+
+// In several of these auto-generated fields, we lie to zod about the output
+// type of the field - with type assertions, we say that they are non-nullable.
+// This will be true in the database, even though the transform functions don't
+// always return values (e.g. createdTimestamp will always be a Date, it'll just
+// get set on insertion and not on subsequent updates).
+
+export const stringId = z.string().regex(Id).optional()
+  .transform((v) => v as unknown as string);
+(stringId as any).customJsonSchema = { bsonType: 'string', pattern: Id.source };
+
+export const createdTimestamp = z.date().optional()
+  .transform((v) => {
+    if (v) return v;
+    if (IsInsert.getOrNullIfOutsideFiber() || IsUpsert.getOrNullIfOutsideFiber()) return new Date();
+    return undefined as unknown as Date;
```

Comment:
Does reaching this `undefined` here indicate a programming error?  Should we just throw instead?

Reply:
No, it indicates that we're applying the schema in a non-insert, non-upsert (i.e. update) context. That means that we're only updating a subset of the schema (because we don't support whole-document updates), so it's OK if this field remains unpopulated.